src/cargo/ops/cargo_rustc/context/mod.rs

```diff
@@ -770,236 +781,25 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
-                        return false;
-                    }
-
-                    // If we've gotten past all that, then this dependency is
-                    // actually used!
-                    true
-                })
-        }).filter_map(|id| match self.get_package(id) {
-                Ok(pkg) => pkg.targets().iter().find(|t| t.is_lib()).map(|t| {
-                    let unit = Unit {
-                        pkg,
-                        target: t,
-                        profile: self.lib_or_check_profile(unit, t),
-                        kind: unit.kind.for_target(t),
-                    };
-                    Ok(unit)
-                }),
-                Err(e) => Some(Err(e)),
-            })
-            .collect::<CargoResult<Vec<_>>>()?;
-
-        // If this target is a build script, then what we've collected so far is
-        // all we need. If this isn't a build script, then it depends on the
-        // build script if there is one.
-        if unit.target.is_custom_build() {
-            return Ok(ret);
-        }
-        ret.extend(self.dep_build_script(unit));
-
-        // If this target is a binary, test, example, etc, then it depends on
-        // the library of the same package. The call to `resolve.deps` above
-        // didn't include `pkg` in the return values, so we need to special case
-        // it here and see if we need to push `(pkg, pkg_lib_target)`.
-        if unit.target.is_lib() && !unit.profile.doc {
-            return Ok(ret);
-        }
-        ret.extend(self.maybe_lib(unit));
-
-        // Integration tests/benchmarks require binaries to be built
-        if unit.profile.test && (unit.target.is_test() || unit.target.is_bench()) {
-            ret.extend(
-                unit.pkg
-                    .targets()
-                    .iter()
-                    .filter(|t| {
-                        let no_required_features = Vec::new();
-
-                        t.is_bin() &&
-                // Skip binaries with required features that have not been selected.
-                t.required_features().unwrap_or(&no_required_features).iter().all(|f| {
-                    self.resolve.features(id).contains(f)
-                })
-                    })
-                    .map(|t| Unit {
-                        pkg: unit.pkg,
-                        target: t,
-                        profile: self.lib_or_check_profile(unit, t),
-                        kind: unit.kind.for_target(t),
-                    }),
-            );
-        }
-        Ok(ret)
-    }
-
-    /// Returns the dependencies needed to run a build script.
-    ///
-    /// The `unit` provided must represent an execution of a build script, and
-    /// the returned set of units must all be run before `unit` is run.
-    pub fn dep_run_custom_build(&self, unit: &Unit<'a>) -> CargoResult<Vec<Unit<'a>>> {
+    // TODO: this ideally should be `-> &[Unit<'a>]`
```

Comment:
Is this not possible due to various methods that take `&mut self` on `Context`?

Reply:
Yep! Specifically, this happens because `target_filenames` and `target_metadatas` are calculated lazily. I hope to refactor them out as well, and change a whole bunch of methods from `&mut` to `&`.